speechtotextv1/websocket_listener.go

```diff
@@ -0,0 +1,144 @@
+package speechtotextv1
+
+import (
+	"encoding/json"
+	"github.com/gorilla/websocket"
+	core "github.com/watson-developer-cloud/go-sdk/core"
+	"io"
+	"time"
+)
+
+type websocketListener struct {
+	RespChannel chan *core.DetailedResponse
+	WSOpenChan  chan *websocket.Conn
+	IsClosed    chan bool
+	Callback    RecognizeCallbackWrapper
+	IsListening bool
+}
+
+/*
+	OnOpen: Sends start message to server when connection created
+*/
+func (wsHandle websocketListener) OnOpen(recognizeOpt *RecognizeOptions) {
+	conn := <-wsHandle.WSOpenChan
+	wsHandle.Callback.OnOpen()
+	sendStartMessage(conn, recognizeOpt)
+}
+
+/*
+	OnClose: Callback when websocket connection is closed
+*/
+func (wsHandle websocketListener) OnClose(finish chan bool) {
+	<-wsHandle.IsClosed
+	wsHandle.Callback.OnClose()
+	finish <- true
+}
+
+/*
+	OnData: Callback when websocket connection receives data
+*/
+func (wsHandle websocketListener) OnData(conn *websocket.Conn, recognizeOptions *RecognizeOptions) {
+	for {
+		var websocketResponse SpeechRecognitionResults
+		_, result, err := conn.ReadMessage()
+		if err != nil {
+			wsHandle.OnError(err)
+		}
+		json.Unmarshal(result, &websocketResponse)
+
+		if websocketResponse.State == "listening" {
```

Comment:
We wouldn't need the `isListening` logic like python. Could you refer to this example: https://github.com/gorilla/websocket/blob/master/examples/echo/client.go#L40

We would essentially need to stream the data to the service in a go routine.